modifications/DataClass.cc

```diff
@@ -132,34 +148,79 @@ void DataClass::uploadKeyToServer(QString location, QString pathOfKey)
     QByteArray postData = doc.toJson();
 
     QNetworkReply* reply = manager.post(request, postData);
-    connect(reply, &QNetworkReply::finished, this, &DataClass::readyReadPublicKey);
-
+    connect(reply, &QNetworkReply::finished, this, &DataClass::readyReadUploadDronePublicKeyToServer);
 }
 
-void DataClass::readyReadPublicKey()
+void DataClass::readyReadUploadDronePublicKeyToServer()
 {
     QNetworkReply* reply = qobject_cast<QNetworkReply*>(sender());
     QString replyStr = reply->readAll();
     int statusCode = reply->attribute( QNetworkRequest::HttpStatusCodeAttribute).toInt();
     if(statusCode == 201 || statusCode == 200){
         emit keyUploadSuccessful();
+        getServerPublicKey();
     }else{
+        showMessage("Unable to Upload Drone Public Key to Server", statusCode);
         emit keyUploadFailed();
     }
+}
+//*******************************************************************************************************************************************************************************
+//SERVER PUBLIC KEY ROTATION
+void DataClass::getServerPublicKey()
+{
+    QString location = oAuthServerPublicKeyUrl;
+    QNetworkRequest request = QNetworkRequest(location);
+    request.setRawHeader("Authorization",QByteArray("Bearer ").append(accessToken));
+    QNetworkReply* reply = manager.get(request);
+    connect(reply,&QNetworkReply::finished, this, &DataClass::readyReadGetServerPublicKey);
+}
 
+void DataClass::readyReadGetServerPublicKey()
+{
+    QNetworkReply* reply = qobject_cast<QNetworkReply*>(sender());
+    QString replyStr = reply->readAll();
+    int statusCode = reply->attribute( QNetworkRequest::HttpStatusCodeAttribute).toInt();
+    if(statusCode == 201 || statusCode == 200){
+        QJsonDocument jsonReply = QJsonDocument::fromJson(replyStr.toUtf8());
+        QJsonArray jsonArray = jsonReply.object()["keys"].toArray();
+        if(jsonArray.size() > 0){
+            QJsonObject jwks = jsonArray[0].toObject();
+            std::string e = jwks["e"].toString().toStdString();
+            std::string kid = jwks["kid"].toString().toStdString();
+            std::string n = jwks["n"].toString().toStdString();
+            if(!SignatureVerifier::convertJwkToPem(n,e,kid,serverPublicKey)){
+                emit serverPublicKeyDownloadFailed();
+                return;
+            }
+            emit serverPublicKeyDownloadSuccessful();
+            // GET ALL THE DATA FROM SERVER
+            getAllActivities();
```

Comment:
Maybe have a better name, one would expect get server public key method to just get server public key. You seem to be doing a lot more here. or does the ready prefix mean after getServerPublicKey is done?

Reply:
readyRead prefix is used when the reply of the API request is recieved.